### ISSUE TITLE
Unable to push image to DockerHub 

### ISSUE THREAD

**Issue by rvichery** [openshift/origin#17064]
I followed this blog post: [https://blog.openshift.com/pushing-application-images-to-an-external-registry/](https://blog.openshift.com/pushing-application-images-to-an-external-registry/) but I am unable to push a container image to DockerHub from OpenShift. I am getting the following error at the end of the build:
```
error: build error: Failed to push image: unauthorized: authentication required
```

I was able to workaround it by modifying the dockerhub secret from:
```
{
        "auths": {
                "https://index.docker.io/v1": {
                        "auth": "XXXXXXXXXXXXX",
                        "email": "XXXXXXXXXXXXX"
                 }
        }
}
```
to
```
{
        "auths": {
                "docker.io": {
                        "auth": "XXXXXXXXXXXXX",
                        "email": "XXXXXXXXXXXXX"
                 }
        }
}
```

I tried to change the output specs in the buildconfig from docker.io/rvichery/myproject:latest to index.docker.io/rvichery/myproject:latest to see if this was causing the issue, and I got the same result. It seems that the entity pushing the container is doing a strict match on the registry name. If I put index.docker.io and not https://index.docker.io/v1 in the dockerhub secret and prefix the container image with index.docker.io as well it works. 

##### Version
oc v3.6.173.0.21
kubernetes v1.6.1+5115d708d7
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://openshift-master-rhel.bdteam.local:8443
openshift v3.6.173.0.21
kubernetes v1.6.1+5115d708d7

##### Steps To Reproduce
1. docker login docker.io
2. oc secrets dockerhub  .docker/config.json
3. oc start-build

##### Current Result
Container image is not being pushed to DockerHub

##### Expected Result
Container image pushed to DockerHub

##### Additional Information
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
kind: BuildConfig
metadata:
  annotations:
    openshift.io/generated-by: OpenShiftNewBuild
  creationTimestamp: 2017-10-26T23:46:51Z
  labels:
    build: time
  name: time
  namespace: gitlab
  resourceVersion: "77741"
  selfLink: /oapi/v1/namespaces/gitlab/buildconfigs/time
  uid: f0b3aed5-baa7-11e7-b186-06d1b61fb4f0
spec:
  failedBuildsHistoryLimit: 5
  nodeSelector: null
  output:
    pushSecret:
      name: dockerhub
    to:
      kind: DockerImage
      name: docker.io/rvichery/myproject:latest
  postCommit: {}
  resources: {}
  runPolicy: Serial
  source:
    contextDir: busybox
    git:
      uri: https://github.com/rvichery/myproject
    type: Git
  strategy:
    dockerStrategy:
      from:
        kind: DockerImage
        name: busybox:latest
    type: Docker
  successfulBuildsHistoryLimit: 5
  triggers:
  - github:
      secret: XXXXXXXXXXXXXXXXX
    type: GitHub
  - generic:
      secret: XXXXXXXXXXXXXXXXX
    type: Generic
  - type: ConfigChange
status:
  lastVersion: 16
```

**Comment by xiaods** [comment]
out of date. need refresh the docs.

**Comment by pweil-** [comment]
Thank you for reporting this issue with a blog post.  I will notify the author to request it be updated.

**Comment by VeerMuchandi** [comment]
Tried it today. It works well for me

$ oc logs -f time-5-build
Step 1 : FROM busybox@sha256:91ef6c1c52b166be02645b8efee30d1ee65362024f7da41c404681561734c465
 ---> 6ad733544a63
Step 2 : MAINTAINER Veer Muchandi veer@redhat.com
 ---> Using cache
 ---> 3cad3654c221
Step 3 : ADD ./init.sh ./
 ---> Using cache
 ---> f2e77cf00c1c
Step 4 : EXPOSE 8080
 ---> Using cache
 ---> 7231254fd6b8
Step 5 : CMD ./init.sh
 ---> Using cache
 ---> ddd4292eb946
Step 6 : ENV "OPENSHIFT_BUILD_NAME" "time-5" "OPENSHIFT_BUILD_NAMESPACE" "ext-image-push" "OPENSHIFT_BUILD_SOURCE" "https://github.com/VeerMuchandi/time" "OPENSHIFT_BUILD_COMMIT" "39a16e6b59ad96b0a15714e54d6eaebbaa1fb164"
 ---> Running in dee1a1648b84
 ---> 3a3ddccc6c6c
Removing intermediate container dee1a1648b84
Step 7 : LABEL "io.openshift.build.commit.date" "Wed Sep 14 20:45:02 2016 -0400" "io.openshift.build.commit.message" "Update init.sh" "io.openshift.build.commit.author" "VeerMuchandi \u003cveer.muchandi@gmail.com\u003e" "io.openshift.build.commit.id" "39a16e6b59ad96b0a15714e54d6eaebbaa1fb164" "io.openshift.build.commit.ref" "master" "io.openshift.build.source-location" "https://github.com/VeerMuchandi/time" "io.openshift.build.source-context-dir" "busybox" "io.openshift.build.name" "time-5" "io.openshift.build.namespace" "ext-image-push"
 ---> Running in 698f6f9b4d53
 ---> a6d15e9952c9
Removing intermediate container 698f6f9b4d53
Successfully built a6d15e9952c9
Pushing image docker.io/veermuchandi/mytime:latest ...
Pushed 0/2 layers, 32% complete
Pushed 1/2 layers, 100% complete
Pushed 2/2 layers, 100% complete
Push successful

Build Configuration still has the same changes
  output:
    pushSecret:
      name: dockerhub
    to:
      kind: DockerImage
      name: docker.io/veermuchandi/mytime:latest


I am using oc 3.7.9 though

$ oc version
oc v3.7.9
kubernetes v1.7.6+a08f5eeb62
features: Basic-Auth

Server https://master.devday.ocpcloud.com:443
openshift v3.7.9
kubernetes v1.7.6+a08f5eeb62

One thing I noticed is that I had to disable "Securely store docker logins in MacOS keychain" in the docker preferences if you are using a Mac. With this enabled the auth keys are not directly stored in the config.json file and build would throw an exception.
